File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/messages/CredentialOfferMessage.java
package msdk.java.messages;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

import msdk.java.utils.CommonUtils;

public class CredentialOfferMessage {
    public String name;
    public JSONObject attributes;
    public String offer;
    public String threadId;

    public CredentialOfferMessage(String name, JSONObject attributes, String offer, String threadId) {
        this.name = name;
        this.attributes = attributes;
        this.offer = offer;
        this.threadId = threadId;
    }

    public static CredentialOfferMessage parse(String message) {
        try {
            JSONObject json = new JSONObject(message);
            String threadId = CommonUtils.getThreadId(json);
            String name = json.getString("comment");
            JSONObject attributes = extractAttributesFromCredentialOffer(json);
            return new CredentialOfferMessage(
                    name,
                    attributes,
                    message,
                    threadId
            );
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static JSONObject extractAttributesFromCredentialOffer(JSONObject offer) {
        JSONObject attributes = new JSONObject();
        try {
            JSONArray previewAttributes = offer.getJSONObject("credential_preview").getJSONArray("attributes");

            for (int i = 0; i < previewAttributes.length(); i++) {
                JSONObject attribute = previewAttributes.getJSONObject(i);
                attributes.put(attribute.getString("name"), attribute.getString("value"));
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return attributes;
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/messages/CredentialOfferMessage.kt
package msdk.kotlin.sample.messages

import msdk.kotlin.sample.utils.CommonUtils
import org.json.JSONException
import org.json.JSONObject


class CredentialOfferMessage(
    var name: String,
    var attributes: JSONObject,
    var offer: String?,
    var threadId: String
) {

    companion object {
        fun parse(message: String?): CredentialOfferMessage? {
            return try {
                val json = JSONObject(message)
                val threadId: String = CommonUtils.getThreadId(json)!!
                val name = json.getString("comment")
                val attributes = extractAttributesFromCredentialOffer(json)
                CredentialOfferMessage(
                    name,
                    attributes!!,
                    message,
                    threadId
                )
            } catch (e: JSONException) {
                e.printStackTrace()
                null
            }
        }

        fun extractAttributesFromCredentialOffer(offer: JSONObject): JSONObject? {
            val attributes = JSONObject()
            try {
                val previewAttributes =
                    offer.getJSONObject("credential_preview").getJSONArray("attributes")
                for (i in 0 until previewAttributes.length()) {
                    val attribute = previewAttributes.getJSONObject(i)
                    attributes.put(attribute.getString("name"), attribute.getString("value"))
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return attributes
        }
    }
}